lib/spack/spack/binary_distribution.py

```diff
@@ -37,23 +39,294 @@
+    specs available on a mirror)
+
+    2. a cache of all the conrcrete built specs available on all the
+    configured mirrors.
+    """
+
+    def __init__(self, cache_root=None):
+        self._cache_root = cache_root or default_manager_root
+        self._index_cache_root = os.path.join(self._cache_root, 'indices')
+        self._index_contents_key = 'contents.json'
+        self._index_file_cache = None
+        self._local_index_cache = None
+        self._built_spec_cache = {}
+
+    def _init_local_index_cache(self):
+        if not self._index_file_cache:
+            self._index_file_cache = file_cache.FileCache(
+                self._index_cache_root)
+
+            cache_key = self._index_contents_key
+            self._index_file_cache.init_entry(cache_key)
+
+            cache_path = self._index_file_cache.cache_path(cache_key)
+
+            self._local_index_cache = {}
+            if os.path.isfile(cache_path):
+                with self._index_file_cache.read_transaction(
+                        cache_key) as cache_file:
+                    self._local_index_cache = json.load(cache_file)
+
+    @property
+    def spec_cache(self):
+        """The cache of specs and mirrors they live on"""
+        return self._built_spec_cache
+
+    def clear_spec_cache(self):
+        self._built_spec_cache = {}
+
+    def _write_local_index_cache(self):
+        self._init_local_index_cache()
+        cache_key = self._index_contents_key
+        with self._index_file_cache.write_transaction(cache_key) as (old, new):
+            json.dump(self._local_index_cache, new)
+
+    def regenerate_spec_cache(self):
+        self._init_local_index_cache()
+
+        for mirror_url in self._local_index_cache:
+            cache_entry = self._local_index_cache[mirror_url]
+            cached_index_path = cache_entry['index_path']
+            self._associate_built_specs_with_mirror(cached_index_path,
+                                                    mirror_url)
+
+    def _associate_built_specs_with_mirror(self, cache_key, mirror_url):
+        self._init_local_index_cache()
+
+        tmpdir = tempfile.mkdtemp()
+
+        try:
+            db_root_dir = os.path.join(tmpdir, 'db_root')
+            db = spack_db.Database(None, db_dir=db_root_dir,
+                                   enable_transaction_locking=False)
+
+            self._index_file_cache.init_entry(cache_key)
+            cache_path = self._index_file_cache.cache_path(cache_key)
+            with self._index_file_cache.read_transaction(cache_key):
+                db._read_from_file(cache_path)
+
+            spec_list = db.query_local(installed=False)
+
+            for indexed_spec in spec_list:
+                indexed_spec_dag_hash = indexed_spec.dag_hash()
+                if indexed_spec_dag_hash not in self._built_spec_cache:
+                    self._built_spec_cache[indexed_spec_dag_hash] = []
+                for entry in self._built_spec_cache[indexed_spec_dag_hash]:
```

Comment:
I think my worry was that if `binary_distribution.get_specs` was called multiple times, I didn't want to append duplicate entries to the end of the list. But in that case, now that you point it out, I wonder if (when I encounter an entry with the same `mirror_url`) I should replace the `spec` key in that entry instead of just breaking.  When a subsequent call to `get_specs` results in a DAG hash match at a certain mirror, it seems possible it could actually be a new binary (with a different `full_hash`).